---
title: "R Notebook"
output:
  html_document:
    df_print: paged
  html_notebook:
    fig_height: 8
    fig_width: 10
---
de novo transcriptome assemblies have helped for studying non-model organisms for which a sequenced references genome is unavailable. de novo transcript assemblies also can help for discovering novo transcripts, alternative splicing for model based species, organism, or specific cell types. Here we will explore the de novo assembly and establish some analysis strategies. 

First source function file
```{r}
source('~/Works/github/ScDna/R/functions.R')
```
We have used `Trinity` to de novo assemble [708 single cells](https://www.sciencedirect.com/science/article/pii/S0896627317306839). Traditionally, `Blast` would be used to identify de novo assembled transcripts with model genome. Here we use minimap2 instead to map the assembled transcripts back to human genome+transcriptome and return value `1` if minimap2 has a hit on reference genome and `0` if no hit on reference genome. Then, to make analysis easier and faster, we only focus on the `protein_coding` transcripts.

# De Novo Assembled Transcripts
```{r}
cnt<-read.csv('~/Documents/trinity/Docs/idxstat_T_all_protein_coding.csv',header=T)
```
Load meta information about cells
```{r}
meta<-read.csv('~/Documents/HCA/benchmarking/Sra_meta_groups_iPSC_730.csv')
#'  filter out bulk samples
meta.sc<-subset(meta,meta$cell !="population")
label<-paste(meta.sc$cell,meta.sc$lineage,sep='-')
meta.sc['label']<-label
```
remove bulk sequencing samples from assembled data
```{r}
cnt.dd<-cnt[,-c(1:2)]
cnt.hh<-colnames(cnt.dd)
mlist<-match(meta.sc$sra,cnt.hh)
cnt.dd<-cnt.dd[,mlist]
# convert greater than 1 value to 1
cnt.dd[cnt.dd>1] <- 1
```

# Hamming Distance
calculate hamming distance
```{r}
x<-as.matrix(cnt.dd)
hd<- hamming_binary(x)
nn<-nrow(x)
hd.r<-hd/nn
hd.log<-log(hd+1,base=2)
```

# tSNE visualization


Use scaled hamming distance
```{r}
X<- hd.log
```
Run tSNE to visualize matrix
```{r}
hd.tsne<-Rtsne(X,check_duplicates = FALSE)
df<-data.frame("X1"=hd.tsne$Y[,1],"X2"=hd.tsne$Y[,2],meta.sc)
labels<-paste(df$cell,df$lineage)
df$labels<-labels
p.raw<-ggscatter(
  df,
  x = "X1",
  y = "X2",
  color = "labels",
  palette = "ucscgb",
  size = 3,
  alpha = 0.9,
  caption= "tSNE: de novo assembled transcripts",
  ggtheme = theme_minimal()
) + border()+
  xlab('tSNE 1')+
  ylab('tSNE 2')
```
# Clustering

## SNN-Cliq clusering

First, we used `SNN` to cluster de novo assembled transcripts

```{r}
hd.grps <- buildSNNGraph(X)
hd.clusters <- cluster_fast_greedy(hd.grps)
```
project clustering membership onto tSNE plot

```{r fig.height=10, fig.width=20, message=FALSE, warning=FALSE, paged.print=FALSE}
hd.fc <-  membership(hd.clusters)
df$membership<-as.factor(hd.fc)
print(p.raw)
p.snn<-ggscatter(
  df,
  x = "X1",
  y = "X2",
  color = "membership",
  palette = "ucscgb",
  size = 3,
  alpha = 0.9,
  caption= "tSNE: SNN clustering of de novo assembled transcripts",
  ggtheme = theme_minimal()
) + border()+
  xlab('tSNE 1')+
  ylab('tSNE 2')
figure <- ggarrange(p.raw, p.snn,
                    labels = c("Cell Labels", "SNN Clustering"),
                    ncol = 2, nrow = 1)
figure
```
## kNN Louvain

```{r fig.height=10, fig.width=20, message=FALSE, warning=FALSE, paged.print=FALSE}
out<-plotTSneClustering(X,hd.tsne,'Louvain')
figure <- ggarrange(p.raw, out$pl,
                    labels = c("Cell Labels", "KNN Louvain Cluster"),
                    ncol = 2, nrow = 1)
figure
```
